Add check for missing problem-specifications directory

nywilken opened the pull request:
Readme generation depends on the existence of the problem-specifications repository. This change fixes an issue where executing a readme generation with an zero-value ProblemSpecification results in a incomplete Readme.md file. 

The configlet command will now print an error message if it is unable to load the problem-specifications repository.

Closes exercism/configlet#27

nywilken commented:
@kytrinyx @robphoenix I saw this issue and took the opportunity to divide into how readme generation works a little further. I plan to add a test or two to verify error handling but, as always, I'm open to initial thoughts and feedback.

kytrinyx commented on cmd/generate.go:
```diff
@@ -62,8 +62,12 @@ func generate(cmd *cobra.Command, args []string) {
 			continue
 		}
+
+		// TODO find the best way to handle multiple errors.
```
There's a multi errors library from hashicorp that I've used successfully before.

nywilken replied:
Sweet, gurssing you referring to https://github.com/hashicorp/go-multierror

kytrinyx replied:
Yeah! That's the one. Thanks for finding it—I've had a very confusing day and got distracted before I remembered to do that.

kytrinyx commented on track/exercise_readme.go:
```diff
@@ -40,16 +40,15 @@ func NewExerciseReadme(root, trackID, slug string) (ExerciseReadme, error) {
 	}
 
+	// Don't care if this errors here as errors are handled, where it counts, in readme.Generate().
+	if err := readme.loadProblemSpecification(); err != nil {
```
Do we load the problem specification twice, once here and once in Generate?

nywilken replied:
So problem specification is only loaded once, but the call is made twice to support the loading of the problem specification upon ExerciseReadme construction. I had a version of the code without loading problem specification in the constructor - only within readme.Generate() -  but added it back once I saw that readme.Spec would remain nil until calling readme.Generate(). 

Was the original intent to not have ProblemSpecification be a hard dependency for the initialization of an ExerciseReadme? In other words, it should not error if the the loading of the problem specification failed? 

I ask because, I thought to error here but the original code sort of swallowed the error and set it to an empty ProblemSpecification. What do you think about making it a hard dependency, if problem specification fails to load return immediately like we do for readTemplate()?

kytrinyx replied:
I'm fine with making it a hard dependency. I think that making it a not-hard dependency was accidentaly.

kytrinyx replied:
Oh, also: this would mean that we don't need a specific "can't find the problem spec dir" test, though that might be easier to debug than an error coming from farther down. I don't know though—if the error message is "can't find a problem spec" then it might well be just as easy to see what's up.

nywilken replied:
Great, thanks for your thoughts here. I have a working version that I hope to push up either tonight or in the am - been mobile pretty much the whole day. 

As for where the error message should live, I think in both places is probably the best for correctness. 

In the generate command, if spec-path doesn't exist we should error with a friendly message "hmmm, seems like the problem specification path provided doesn't exist; check the path and try again". This would exit immediately as there is no point in trying to initialize an exercise readme if we know it will fail. 

In the case where spec-path does exist, the check within the the exercise readme will error if it can't read the problem specification data for one, more or all exercises. Preventing the generation of an incomplete readme. 

Sounds good to you?

kytrinyx replied:
Yepp, that sounds good to me!

kytrinyx commented on track/exercise_readme.go:
```diff
@@ -63,6 +62,10 @@ func NewExerciseReadme(root, trackID, slug string) (ExerciseReadme, error) {
+	if err := readme.loadProblemSpecification(); err != nil {
+		return "", err
+	}
+
```
This is loading it specifically so that it can check the error, right?

Instead of doing that, maybe we can have a specific test at the command level that says "problem specification directory exists".

I guess that would make it possible to have an exercise that is missing metadata entirely, but that feels like a different problem (linting, perhaps).

nywilken replied:
> This is loading it specifically so that it can check the error, right?
Yes.

> Instead of doing that, maybe we can have a specific test at the command level that says "problem specification directory exists".

I agree with this approach. I left a comment, above, about my first implementation that I would like to get your thoughts on.

kytrinyx commented on track/exercise_readme.go:
```diff
@@ -82,6 +85,26 @@ func (readme ExerciseReadme) Write() error {
 
+// loadProblemSpecification locates Problem Specification repository data and initializes Spec for readme.
+// If Spec for readme has already been initialized loadProblemSpecification does nothing.
+func (readme *ExerciseReadme) loadProblemSpecification() error {
```
Oh, I see. Right. Hm.

I think I still would like to be explicit about checking for the existence of the problem-specifications repository up front rather than burying this down here with extra caching.

nywilken replied:
I agree, explicit is the way to go. Thanks for reviewing.

kytrinyx commented on cmd/generate.go:
```diff
@@ -32,11 +33,20 @@ var (
 	path, err := filepath.Abs(filepath.FromSlash(args[0]))
 	if err != nil {
-		fmt.Fprintf(os.Stderr, err.Error())
+		fmt.Println(err.Error())
```
Is it intentional to put this onto stdout instead of stderr?

(Side note, I'm pretty sure you can say `fmt.Println(err)` rather than `fmt.Println(err.Error())`)

nywilken replied:
> Is it intentional to put this onto stdout instead of stderr?

Yeah, I realized when testing the multi error output that I couldn't pipe the error messages to less without redirecting `... 2>&1` so I  changed to use Stodout. 

But maybe the correct thing to do here is implement a function for displaying multiple errors with a different level of verbosity. Or add a verbosity flag. I'm open to suggestions. 

> Side note ...

Good catch, I moved from Fprintf but didn't change the argument.

nywilken replied:
> so I changed to use Stodout.

In hindsight, this should probably be it's own PR so we can discuss Stderr vs Stdout for errors. I will break out this commit into its own PR.

kytrinyx replied:
Thanks, yes, I'd like to discuss this.

kytrinyx commented on cmd/generate.go:
```diff
@@ -32,11 +33,20 @@ var (
+	if specPath != "" {
+		_, err := os.Stat(specPath)
+		if os.IsNotExist(err) {
+			fmt.Printf("hmm, it looks like %s doesn't exist; please check the path and try again.", specPath)
```
I'm wondering if it's worth shortening this a bit, especially since the path is likely to be quite long. Maybe just:

> Path %s not found.

I suppose best practice says that we should suggest how to fix the error, but in this case I think that "not found" kind of obviously means "check that it's actually there".

nywilken replied:
I'm with you on making this shorter 🙌🏽
I had to resize my tmux pane a few times just to read 😂  

`Path Not Found` is perfect.

nywilken replied:
Sorry to nit pick here but seeing as we are displaying an error to the user, similar to the log call in the example, and not returning it to some other caller I think it may be okay to format the error message a little differently. 

What do you think about using something similar to what we do in the lint command  `-> The path '/for/bar' does not exist.`?

kytrinyx replied:
Yeah, we should probably have a discussion at some point about what we do with errors when they are the last stop and we know we're going to show them to the user.

kytrinyx commented on cmd/generate.go:
```diff
@@ -32,11 +33,20 @@ var (
 	}
 
-	track.ProblemSpecificationsPath = specPath
+	if specPath != "" {
```
Thinking out loud, let's set the specPath if it's not empty, then check the existence of `track.ProblemSpecificationsPath`, afterwards.

This will catch the problem in both cases (whether they've specified a custom location, or just using the default location).

nywilken replied:
ProblemSpecificationsPath is an empty string by default and ProblemSpecification will only use it if it is set to anything other than empty string. 

Which makes sense seeing as it at least needs to know some root path, which is not known until initializing. 

Let me think about this a little to see if we can refactor and if it makes sense to.

kytrinyx replied:
> ProblemSpecificationsPath is an empty string by default and ProblemSpecification will only use it if it is set to anything other than empty string.

How about we have an `init` in the `track` package that sets the ProblemSpecPath to the default?

nywilken replied:
That would be ideally, but we need to know at least the root path to the track, before calling init which wouldn't be possible without running the command with some track path, right? 

So instead I added a constant `ProblemSpecificationsDir` to the track package that can be used in the generate command to build the default ProblemSpecificationsPath based off the provided track path. 

```
track.ProblemSpecificationsPath = filepath.Join(root, track.ProblemSpecificationsDir)
// check if track.ProblemSpecificationsPath exists; if no error out 
...
```

I will push up for testing and feedback.

nywilken replied:
Oh a quick question, was there some plans on storing the ProblemSpecPath in the ProblemSpecification type? I see https://github.com/exercism/configlet/blob/master/track/problem_specification.go#L32 but it is not used nor set.

nywilken commented:
@kytrinyx @robphoenix I reverted the changes for printing errors and changed the logic a bit for ProblemSpecificationsPath. Let me know what you think, thanks.

robphoenix commented on cmd/generate.go:
```diff
@@ -35,13 +36,20 @@ func generate(cmd *cobra.Command, args []string) {
+	}
 
+	if _, err := os.Stat(track.ProblemSpecificationsPath); os.IsNotExist(err) {
+		fmt.Fprintf(os.Stderr, "path '%s' not found\n", track.ProblemSpecificationsPath)
```
a little nitpick; with this error message I think the reason should be separate from the cause, ie. `path not found: %s` rather than the path within the error message. This seems to be more idiomatic, and also is easier to follow I think, as well as if the path is long then the `not found` part could be harder to notice.

kytrinyx replied:
I think `path not found: %s` is an excellent call.

We should still have that discussion about user-consumed messages eventually, though!

nywilken replied:
This is a simple enough change so I will update to `path not found: %s`, before merging. Thanks 🙏 

FYI, I spent a little time reading up on the Stderr vs Stdout and came across this great post https://www.jstorimer.com/blogs/workingwithcode/7766119-when-to-use-stderr-instead-of-stdout

After reading, I think deferring to Doug McIlory is the way to go. So I'm going to rescind my Stdout PR, before even creating it.

kytrinyx replied:
> After reading, I think deferring to Doug McIlory is the way to go. So I'm going to rescind my Stdout PR, before even creating it.

❤️

kytrinyx replied:
> I will update to path not found: %s, before merging. 

Perfect. Merge when you're happy, I'd say.
